add graphql-authz blog post

dimatill opened the pull request.

dimatill commented:
@Urigo I've added some code snippets as we discussed

n1ru4l commented on pages/blog/graphql-authz.mdx:
````diff
@@ -0,0 +1,174 @@
+);
+```
+
+This technique helps to avoid additional calls to data sources by executing authorization logic on top of the entire operation result with the related extra data being populated. Since related data is often stored in the same place it can be fetched from a data source in one call instead of making one call for authorization and one call for actual data populating.
````
```suggestion
By using this technique we reduce remote procedure calls to individual data sources by executing authorization logic on top of GraphQL execution result that is enriched with the additional data specified by the authorization rules. Since related data is often stored in the same place it can be fetched from a data source in one roundtrip (via the GraphQL schema) instead of performing one remote procedure call for the authorization and one call for the actual data populating.
```

n1ru4l commented on pages/blog/graphql-authz.mdx:
```diff
@@ -0,0 +1,174 @@
+
+## Microservices
+
+With GraphQL AuthZ it is possible to implement a centralized gateway authorization layer as well as microservice level authorization. You can choose between storing the whole authorization schema in a holistic way in a gateway independent of GraphQL schema or let GraphQL schema manage authorization as part of itself.
```
```suggestion
With GraphQL AuthZ it is possible to implement a centralized gateway authorization layer as well as microservice level authorization. You can choose between storing the whole authorization schema in a holistic way on a stitched or federated gateway or having dedicated authorization schemas for each subgraph/schema specified by your services. It is even possible to mix both approaches!
```

n1ru4l commented on pages/blog/graphql-authz.mdx:
```diff
@@ -0,0 +1,174 @@
+
+> “Shifting this configuration out of the gateway makes subschemas autonomous, and allows them to push their own configuration up to the gateway—enabling more sophisticated schema releases.” — [schema stitching handbook](https://github.com/gmac/schema-stitching-handbook/tree/master/stitching-directives-sdl)
+
+The `@graphql-authz/directive` package provides an importable directive that can be used to annotate types and fields within subschemas and a configuration transformer that can be used on the gateway to convert the subschema directives into explicit authorization settings.
```
```suggestion
The `@graphql-authz/directive` package provides a GraphQL directive that can be used to annotate types and fields within your subschemas SDL and a configuration transformer that can be used on the gateway to convert the subschema directives into explicit authorization settings.
```

n1ru4l commented on pages/blog/graphql-authz.mdx:
````diff
@@ -0,0 +1,174 @@
+}
+```
+
+Directives could be used in monolith architecture as well.
````
```suggestion
On top of that directives can as well be used in monolith architecture.
```

n1ru4l commented on pages/blog/graphql-authz.mdx:
```diff
@@ -0,0 +1,174 @@
+
+Directives could be used in monolith architecture as well.
+
+If directives are not applicable for some reason, plain JSON auth schema could be used to describe which rules should be executed for certain Object, Interface, Field, Query, or Mutation
```
```suggestion
If you are not pursuing a schema-first (SDL) development flow and are more a fan of the code-first approach (which does not let you specify directives), the authorization schema can also be described as a plain JSON object. Allowing you to specify the rules that should be executed on your object types, interfaces, and fields.
```

dimatill replied:
disagree here about the 
> code-first approach (which does not let you specify directives)

it's possible to specify directives with a code-first approach, there is an example with NestJS: https://github.com/AstrumU/graphql-authz/blob/main/examples/packages/nestjs/src/index.ts#L111

n1ru4l replied:
Interesting, I haven't seen this before. When using graphql-js API for building an executable schema using code this is not possible. So nestjs is definitely doing some framework magic on top of that. Maybe we should change `(which does not let you specify directives)` to `(which does not let you specify directives without framework-specific voodoo magic)`? :D

In another library I have built, I allowed mirroring directive functionality through extension fields see https://github.com/n1ru4l/graphql-public-schema-filter#usage-instructions

dimatill replied:
Yeah, agree, I added this :) 

https://github.com/n1ru4l/graphql-public-schema-filter is a nice tool :) yeah, in graphql-authz I also allowed putting auth configuration right to extensions

n1ru4l commented on pages/blog/graphql-authz.mdx:
````diff
@@ -0,0 +1,174 @@
+};
+```
+
+## Comparison with GraphQL Shield
````
```suggestion
## Comparing GraphQL AuthZ with GraphQL Shield
```

n1ru4l commented on pages/blog/graphql-authz.mdx:
```diff
@@ -0,0 +1,174 @@
+
+## Comparison with GraphQL Shield
+
+[GraphQL Shield](https://github.com/maticzav/graphql-shield) is a great tool for creating authorization layers. Actually, GraphQL AuthZ was inspired by GraphQL Shield! However, it incorporates a different approach compared to GraphQL AuthZ in terms of hooking up authorization logic to the execution process.
```
```suggestion
[GraphQL Shield](https://github.com/maticzav/graphql-shield) is a great tool for creating authorization layers that has vast adoption from the community. In fact, GraphQL AuthZ is highly inspired by GraphQL Shield! However, GraphQL Shield uses a different approach compared to GraphQL AuthZ for applying authorization rules.
```

n1ru4l commented on pages/blog/graphql-authz.mdx:
```diff
@@ -0,0 +1,174 @@
+
+[GraphQL Shield](https://github.com/maticzav/graphql-shield) is a great tool for creating authorization layers. Actually, GraphQL AuthZ was inspired by GraphQL Shield! However, it incorporates a different approach compared to GraphQL AuthZ in terms of hooking up authorization logic to the execution process.
+
+The main difference is GraphQL Shield uses field middlewares to execute authorization logic while GraphQL AuthZ wraps the entire execution process. The benefits of the wrapping approach are described in previous paragraphs however there are some drawbacks. For example, with GraphQL AuthZ post-execution rules there is no ability to fail the request earlier because post-execution rules are executed ...
```
```suggestion
The main difference is GraphQL Shield uses field middleware by wrapping all the `resolver` functions within your GraphQL schema for executing authorization logic during the data-resolving phase, while GraphQL AuthZ wraps the entire `execute` logic.

The benefits of the wrapping approach are described in previous paragraphs, however, there are also some drawbacks. For example, with GraphQL AuthZ post-execution rules there is no ability to fail the request early because post-execution rules are executed after all resolvers are executed. GraphQL Shield, on the other hand, can fail the request during the execution phase which happens before the post-execution phase but later than the pre-execution phase. On the other hand, post-execution rules have the benefit of accessing the resolved value of the field they are specified on and, furthermore, even the sibling fields that are specified via the rules selection set.

The middleware approach of GraphQL shield is missing these possibilities because authorization logic is executed in the context of field resolvers and has access only to the parent object which is the value returned from the parent field resolver. At that stage, the wrapped field resolvers have not been executed yet.
```

n1ru4l commented on pages/blog/graphql-authz.mdx:
```diff
@@ -0,0 +1,174 @@
+
+## Conclusion
+
+Please check out the [GraphQL AuthZ GitHub repo](https://github.com/AstrumU/graphql-authz). You can find a tutorial on how to set it up in different ways and with different technologies. Also, you can find the following working examples:
```
```suggestion
GraphQL AuthZ is a new approach to doing GraphQL native authorization. We are happy that we can finally share this library with the community and keen to learn about the ways it might be used within your next project! Don't hesitate to contact us for questions or to evaluate whether this library might be a fit for your architecture.

Please check out the [GraphQL AuthZ repository on GitHub](https://github.com/AstrumU/graphql-authz) for learning more. You can find a tutorial on how to set it up in different ways and with different technologies.

The repository also contains the following, ready to run, examples:
```

n1ru4l reviewed (commented):
Hey, @dimatill I added a few suggestions for wording changes, so stuff is more clear, less redundant, and more engaging in some cases. Let me know what you think 😊

I really like the approach this library takes and can see that this is definitely something we need in the eco-system 🎉😍

dimatill commented:
> Hey, @dimatill I added a few suggestions for wording changes, so stuff is more clear, less redundant, and more engaging in some cases. Let me know what you think 😊
> 
> I really like the approach this library takes and can see that this is definitely something we need in the eco-system 🎉😍

Hi, @n1ru4l it's super awesome, thank you for your feedback and I very much appreciate your suggested changes! I've added you to the authors of this blog post if you're ok with it. 

I've added all the suggestions except one, let's discuss it

n1ru4l commented:
@dimatill I think it would also make sense to add a few paragraphs in the conclusion that sum up use-cases. E.g. when should I use graphql-authz vs doing logic in my resolvers vs graphql-shield vs https://www.envelop.dev/plugins/use-operation-field-permissions

dimatill commented:
> @dimatill I think it would also make sense to add a few paragraphs in the conclusion that sum up use-cases. E.g. when should I use graphql-authz vs doing logic in my resolvers vs graphql-shield vs [envelop.dev/plugins/use-operation-field-permissions](https://www.envelop.dev/plugins/use-operation-field-permissions)

@n1ru4l I added some thoughts about this to the new paragraph `Which approach to choose?` please review it, I believe you'll have suggestions 😊

n1ru4l commented on pages/blog/graphql-authz.mdx:
```diff
@@ -0,0 +1,236 @@
+
+## Which approach to choose?
+
+To sum up use-cases and different authorization approaches, here is the tip:
```
```suggestion
Let's wrap up all the mentioned libraries and the use-cases in which they shine:
```

n1ru4l commented on pages/blog/graphql-authz.mdx:
```diff
@@ -0,0 +1,236 @@
+
+## Conclusion
+
+GraphQL AuthZ is a new approach to doing GraphQL native authorization. We are happy that we can finally share this library with the community and keen to learn about the ways it might be used within your next project! Don't hesitate to contact us for questions or to evaluate whether this library might be a fit for your architecture.
```
```suggestion
GraphQL AuthZ is a new approach for applying GraphQL native authorization. We are happy that we can finally share this library with the community and keen to learn about the ways it might be used within your next project! Don't hesitate to contact us for questions or to evaluate whether this library might be a fit for your architecture.
```

n1ru4l commented:
@dimatill added small adjustment, please apply and rebase those + adjust the date then we are ready to merge 🥳

dimatill commented:
> @dimatill added small adjustment, please apply and rebase those + adjust the date then we are ready to merge 🥳

@n1ru4l Thank you! pushed all the changes
